File: src/models/task-plan.ts
/**
 * Task-Plan 数据模型定义
 * 
 * 描述任务 (Task) 与开发计划 (Development Plan) 的从属关系
 * 一对多关系：一个任务可以关联多个开发计划（方案迭代）
 * 
 * @author 李奥纳多 (🐢 架构师)
 * @created 2026-04-16
 */

// ============================================================================
// 枚举类型定义
// ============================================================================

/** 任务状态枚举 */
export enum TaskStatus {
  PENDING = 'pending',           // 待处理
  IN_PROGRESS = 'in_progress',   // 进行中
  TESTING = 'testing',           // 测试中
  COMPLETED = 'completed',       // 已完成
  FAILED = 'failed',             // 已失败
  BLOCKED = 'blocked'            // 已阻塞
}

/** 任务优先级枚举 */
export enum TaskPriority {
  LOW = 'low',                   // 低
  NORMAL = 'normal',             // 普通
  HIGH = 'high',                 // 高
  CRITICAL = 'critical'          // 紧急
}

/** 计划状态枚举 */
export enum PlanStatus {
  DRAFT = 'draft',               // 草稿
  PENDING_REVIEW = 'pending_review',  // 待审核
  APPROVED = 'approved',         // 已通过
  REJECTED = 'rejected',         // 已拒绝
  IN_PROGRESS = 'in_progress',   // 执行中
  COMPLETED = 'completed',       // 已完成
  ARCHIVED = 'archived'          // 已归档
}

/** 计划步骤状态枚举 */
export enum PlanStepStatus {
  PENDING = 'pending',           // 待执行
  IN_PROGRESS = 'in_progress',   // 执行中
  COMPLETED = 'completed',       // 已完成
  SKIPPED = 'skipped',           // 已跳过
  BLOCKED = 'blocked'            // 已阻塞
}

// ============================================================================
// 核心数据模型
// ============================================================================

/**
 * 任务上下文 - 任务的完整背景信息
 */
export interface TaskContext {
  background: string;                    // 任务背景
  requirements: string[];                // 需求列表
  resources: ResourceLink[];             // 相关资源链接
  acceptanceCriteria: string[];          // 验收标准
  dependencies: string[];                // 依赖任务 ID 列表
  techStack: string[];                   // 技术栈
  notes: string;                         // 备注
  [key: string]: any;                    // 允许扩展字段
}

/**
 * 资源链接
 */
export interface ResourceLink {
  title: string;
  url: string;
  type: 'doc' | 'code' | 'design' | 'reference';
}

/**
 * 子任务
 */
export interface SubTask {
  id: string;
  title: string;
  assignee: string;
  completed: boolean;
  completedAt?: string;
  createdAt: string;
}

/**
 * 任务模型 (Task Model)
 * 
 * 表示一个完整的工作任务，可关联多个开发计划
 */
export interface Task {
  // 基础信息
  id: string;                            // 任务 ID (格式：YYYYMMDD-XXX)
  title: string;                         // 任务标题
  description: string;                   // 任务描述
  priority: TaskPriority;                // 优先级
  
  // 状态信息
  status: TaskStatus;                    // 任务状态
  progress: number;                      // 进度百分比 (0-100)
  
  // 人员信息
  assignee: string;                      // 负责人
  creator: string;                       // 创建人
  reviewers: string[];                   // 审核人列表
  
  // 时间信息
  createdAt: string;                     // 创建时间 (ISO 8601)
  updatedAt: string;                     // 更新时间 (ISO 8601)
  startedAt?: string;                    // 开始时间
  completedAt?: string;                  // 完成时间
  deadline?: string;                     // 截止时间
  
  // 从属关系
  activePlanId?: string;                 // 当前激活的计划 ID
  planIds: string[];                     // 关联的计划 ID 列表 (一对多)
  
  // 任务详情
  context: TaskContext;                  // 任务上下文
  subtasks: SubTask[];                   // 子任务列表
  tags: string[];                        // 标签
  
  // 扩展字段
  metadata?: {
    source?: string;                     // 任务来源
    estimatedHours?: number;             // 预估工时
    actualHours?: number;                // 实际工时
    [key: string]: any;
  };
}

/**
 * 计划步骤 (Plan Step)
 * 
 * 开发计划中的具体执行步骤
 */
export interface PlanStep {
  index: number;                         // 步骤序号
  name: string;                          // 步骤名称
  description: string;                   // 步骤描述
  estimatedHours: number;                // 预估工时
  status: PlanStepStatus;                // 步骤状态
  executor: string;                      // 执行人
  startedAt?: string;                    // 开始时间
  completedAt?: string;                  // 完成时间
  output?: string;                       // 产出物描述
  dependencies: number[];                // 依赖的步骤索引
}

/**
 * 开发计划模型 (Development Plan Model)
 * 
 * 任务的从属对象，描述如何完成任务的详细方案
 * 一个任务可以有多个计划（方案迭代），但同一时间只有一个激活计划
 */
export interface DevelopmentPlan {
  // 基础信息
  id: string;                            // 计划 ID (格式：plan_YYYYMMDDHHMMSS)
  taskId: string;                        // 所属任务 ID (外键)
  title: string;                         // 计划标题
  version: number;                       // 版本号 (从 1 开始)
  
  // 状态信息
  status: PlanStatus;                    // 计划状态
  type: 'frontend' | 'backend' | 'testing' | 'design' | 'research' | 'general';
  
  // 人员信息
  creator: string;                       // 创建人 (通常是李奥纳多)
  reviewer?: string;                     // 审核人 (通常是擎天柱)
  executors: string[];                   // 执行人列表
  
  // 时间信息
  createdAt: string;                     // 创建时间
  reviewedAt?: string;                   // 审核时间
  startedAt?: string;                    // 开始执行时间
  completedAt?: string;                  // 完成时间
  
  // 计划内容
  steps: PlanStep[];                     // 执行步骤列表
  estimatedTotalHours: number;           // 总预估工时
  actualTotalHours?: number;             // 总实际工时
  
  // 审核信息
  reviewComment?: string;                // 审核意见
  rejectionReason?: string;              // 拒绝原因
  
  // 从属关系元数据
  isBasedOn?: string;                    // 基于的计划 ID (版本迭代时)
  supersededBy?: string;                 // 被哪个计划替代
  
  // 扩展字段
  metadata?: {
    riskLevel?: 'low' | 'medium' | 'high';
    confidence?: number;                 // 信心指数 0-1
    alternatives?: string[];             // 备选方案描述
    [key: string]: any;
  };
}

// ============================================================================
// 从属关系接口
// ============================================================================

/**
 * Task-Plan 关系摘要
 * 
 * 用于快速查询任务与其计划的关联关系
 */
export interface TaskPlanRelation {
  taskId: string;
  taskTitle: string;
  taskStatus: TaskStatus;
  activePlanId?: string;
  activePlanStatus?: PlanStatus;
  totalPlans: number;
  approvedPlans: number;
  lastPlanCreatedAt: string;
}

/**
 * 计划执行进度
 */
export interface PlanProgress {
  planId: string;
  taskId: string;
  totalSteps: number;
  completedSteps: number;
  progressPercentage: number;
  currentStepIndex?: number;
  nextStepName?: string;
  estimatedRemainingHours: number;
}

// ============================================================================
// API 请求/响应模型
// ============================================================================

/**
 * 创建任务请求
 */
export interface CreateTaskRequest {
  title: string;
  description?: string;
  priority?: TaskPriority;
  assignee: string;
  deadline?: string;
  context?: Partial<TaskContext>;
  tags?: string[];
}

/**
 * 更新任务请求
 */
export interface UpdateTaskRequest {
  title?: string;
  description?: string;
  priority?: TaskPriority;
  status?: TaskStatus;
  progress?: number;
  assignee?: string;
  deadline?: string;
  context?: Partial<TaskContext>;
  tags?: string[];
}

/**
 * 创建开发计划请求
 */
export interface CreatePlanRequest {
  taskId: string;
  title?: string;
  type: DevelopmentPlan['type'];
  steps: Omit<PlanStep, 'status' | 'startedAt' | 'completedAt'>[];
  metadata?: DevelopmentPlan['metadata'];
}

/**
 * 更新开发计划请求
 */
export interface UpdatePlanRequest {
  title?: string;
  status?: PlanStatus;
  steps?: PlanStep[];
  reviewComment?: string;
  metadata?: DevelopmentPlan['metadata'];
}

/**
 * 审核计划请求
 */
export interface ReviewPlanRequest {
  approved: boolean;
  comment?: string;
  rejectionReason?: string;
}

/**
 * 任务响应 (包含嵌套计划)
 */
export interface TaskResponse extends Task {
  plans?: DevelopmentPlan[];             // 嵌套的计划列表
  activePlan?: DevelopmentPlan;          // 当前激活的计划
}

/**
 * 计划响应 (包含所属任务)
 */
export interface PlanResponse extends DevelopmentPlan {
  task?: Task;                           // 嵌套的所属任务
}

// ============================================================================
// 工具函数类型
// ============================================================================

/**
 * 计划生成器接口
 * 
 * 李奥纳多根据任务类型自动生成计划模板
 */
export interface PlanGenerator {
  generatePlanTemplate(task: Task): Omit<CreatePlanRequest, 'taskId'>;
}

/**
 * 任务 - 计划管理器接口
 */
export interface TaskPlanManager {
  // Task CRUD
  createTask(request: CreateTaskRequest): Promise<Task>;
  getTask(taskId: string, includePlans?: boolean): Promise<TaskResponse>;
  updateTask(taskId: string, request: UpdateTaskRequest): Promise<Task>;
  deleteTask(taskId: string): Promise<void>;
  listTasks(filters?: TaskFilters): Promise<Task[]>;
  
  // Plan CRUD
  createPlan(request: CreatePlanRequest): Promise<DevelopmentPlan>;
  getPlan(planId: string, includeTask?: boolean): Promise<PlanResponse>;
  updatePlan(planId: string, request: UpdatePlanRequest): Promise<DevelopmentPlan>;
  deletePlan(planId: string): Promise<void>;
  listPlans(filters?: PlanFilters): Promise<DevelopmentPlan[]>;
  
  // 从属关系管理
  setActivePlan(taskId: string, planId: string): Promise<Task>;
  reviewPlan(planId: string, request: ReviewPlanRequest): Promise<DevelopmentPlan>;
  getTaskPlanRelation(taskId: string): Promise<TaskPlanRelation>;
  getPlanProgress(planId: string): Promise<PlanProgress>;
  
  // 自动计划生成
  generatePlanForTask(taskId: string): Promise<DevelopmentPlan>;
}

/**
 * 任务查询过滤器
 */
export interface TaskFilters {
  status?: TaskStatus | TaskStatus[];
  priority?: TaskPriority | TaskPriority[];
  assignee?: string | string[];
  creator?: string;
  tags?: string[];
  hasActivePlan?: boolean;
  createdAfter?: string;
  createdBefore?: string;
  deadlineAfter?: string;
  deadlineBefore?: string;
  limit?: number;
  offset?: number;
  sortBy?: 'createdAt' | 'updatedAt' | 'deadline' | 'priority';
  sortOrder?: 'asc' | 'desc';
}

/**
 * 计划查询过滤器
 */
export interface PlanFilters {
  taskId?: string;
  status?: PlanStatus | PlanStatus[];
  type?: DevelopmentPlan['type'] | DevelopmentPlan['type'][];
  creator?: string;
  reviewer?: string;
  createdAfter?: string;
  createdBefore?: string;
  limit?: number;
  offset?: number;
  sortBy?: 'createdAt' | 'reviewedAt' | 'completedAt';
  sortOrder?: 'asc' | 'desc';
}

// ============================================================================
// 导出
// ============================================================================

export {
  TaskStatus,
  TaskPriority,
  PlanStatus,
  PlanStepStatus,
  TaskContext,
  ResourceLink,
  SubTask,
  Task,
  PlanStep,
  DevelopmentPlan,
  TaskPlanRelation,
  PlanProgress,
  CreateTaskRequest,
  UpdateTaskRequest,
  CreatePlanRequest,
  UpdatePlanRequest,
  ReviewPlanRequest,
  TaskResponse,
  PlanResponse,
  PlanGenerator,
  TaskPlanManager,
  TaskFilters,
  PlanFilters
};

File: backend/models/task_plan.py
"""
Task-Plan 数据模型定义

描述任务 (Task) 与开发计划 (Development Plan) 的从属关系
一对多关系：一个任务可以关联多个开发计划（方案迭代）

@author 拉斐尔 (🐢 后端开发)
@created 2026-04-16
"""

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import List, Optional, Dict, Any

from sqlalchemy import (
    Column, String, Integer, Text, DateTime, ForeignKey,
    CheckConstraint, Numeric, Boolean, Index, JSON, Enum as SAEnum
)
from sqlalchemy.orm import relationship, validates
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()


def portable_enum(enum_class):
    return SAEnum(enum_class, native_enum=False, validate_strings=True)

# ============================================================================
# 枚举类型定义
# ============================================================================


class TaskStatusEnum(str, Enum):
    PENDING = 'pending'
    IN_PROGRESS = 'in_progress'
    TESTING = 'testing'
    COMPLETED = 'completed'
    FAILED = 'failed'
    BLOCKED = 'blocked'


class TaskPriorityEnum(str, Enum):
    LOW = 'low'
    NORMAL = 'normal'
    HIGH = 'high'
    CRITICAL = 'critical'


class PlanStatusEnum(str, Enum):
    DRAFT = 'draft'
    PENDING_REVIEW = 'pending_review'
    APPROVED = 'approved'
    REJECTED = 'rejected'
    IN_PROGRESS = 'in_progress'
    COMPLETED = 'completed'
    ARCHIVED = 'archived'


class PlanStepStatusEnum(str, Enum):
    PENDING = 'pending'
    IN_PROGRESS = 'in_progress'
    COMPLETED = 'completed'
    SKIPPED = 'skipped'
    BLOCKED = 'blocked'


class PlanTypeEnum(str, Enum):
    FRONTEND = 'frontend'
    BACKEND = 'backend'
    TESTING = 'testing'
    DESIGN = 'design'
    RESEARCH = 'research'
    GENERAL = 'general'


# ============================================================================
# 核心数据模型
# ============================================================================


class Task(Base):
    """
    任务模型 (Task Model)
    
    表示一个完整的工作任务，可关联多个开发计划
    """
    __tablename__ = 'tasks'

    # 基础信息
    id = Column(String(32), primary_key=True)
    title = Column(String(500), nullable=False)
    description = Column(Text)
    priority = Column(portable_enum(TaskPriorityEnum), nullable=False, default=TaskPriorityEnum.NORMAL)

    # 状态信息
    status = Column(portable_enum(TaskStatusEnum), nullable=False, default=TaskStatusEnum.PENDING)
    progress = Column(Integer, nullable=False, default=0)

    # 人员信息
    assignee = Column(String(100), nullable=False)
    creator = Column(String(100), nullable=False)

    # 时间信息
    created_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)
    started_at = Column(DateTime(timezone=True))
    completed_at = Column(DateTime(timezone=True))
    deadline = Column(DateTime(timezone=True))

    # 从属关系
    active_plan_id = Column(String(64), ForeignKey('plans.id', ondelete='SET NULL'))

    # 扩展字段
    context = Column(JSON, default=dict)
    subtasks = Column(JSON, default=list)
    tags = Column(JSON, default=list)
    extra_metadata = Column('metadata', JSON, default=dict)

    # 关系
    plans = relationship(
        'DevelopmentPlan',
        back_populates='task',
        cascade='all, delete-orphan',
        foreign_keys='DevelopmentPlan.task_id',
    )
    active_plan = relationship('DevelopmentPlan', foreign_keys=[active_plan_id], post_update=True)

    # 约束
    __table_args__ = (
        CheckConstraint('progress >= 0 AND progress <= 100', name='tasks_progress_check'),
        CheckConstraint('deadline IS NULL OR deadline > created_at', name='tasks_deadline_check'),
        Index('idx_tasks_status', 'status'),
        Index('idx_tasks_priority', 'priority'),
        Index('idx_tasks_assignee', 'assignee'),
        Index('idx_tasks_creator', 'creator'),
        Index('idx_tasks_deadline', 'deadline'),
        Index('idx_tasks_created_at', 'created_at'),
        Index('idx_tasks_updated_at', 'updated_at'),
        Index('idx_tasks_active_plan', 'active_plan_id'),
    )

    @validates('progress')
    def validate_progress(self, key, value):
        if value < 0 or value > 100:
            raise ValueError('Progress must be between 0 and 100')
        return value

    def to_dict(self, include_plans: bool = False, include_active_plan: bool = False) -> Dict[str, Any]:
        """转换为字典"""
        data = {
            'id': self.id,
            'title': self.title,
            'description': self.description,
            'priority': self.priority.value if self.priority else None,
            'status': self.status.value if self.status else None,
            'progress': self.progress,
            'assignee': self.assignee,
            'creator': self.creator,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None,
            'started_at': self.started_at.isoformat() if self.started_at else None,
            'completed_at': self.completed_at.isoformat() if self.completed_at else None,
            'deadline': self.deadline.isoformat() if self.deadline else None,
            'active_plan_id': self.active_plan_id,
            'context': self.context or {},
            'subtasks': self.subtasks or [],
            'tags': self.tags or [],
            'metadata': self.extra_metadata or {},
        }

        if include_active_plan and self.active_plan:
            data['activePlan'] = self.active_plan.to_dict()

        if include_plans:
            data['plans'] = [plan.to_dict() for plan in self.plans]

        return data


class DevelopmentPlan(Base):
    """
    开发计划模型 (Development Plan Model)
    
    任务的从属对象，描述如何完成任务的详细方案
    一个任务可以有多个计划（方案迭代），但同一时间只有一个激活计划
    """
    __tablename__ = 'plans'

    # 基础信息
    id = Column(String(64), primary_key=True)
    task_id = Column(String(32), ForeignKey('tasks.id', ondelete='CASCADE'), nullable=False)
    title = Column(String(500), nullable=False)
    version = Column(Integer, nullable=False, default=1)
    type = Column(portable_enum(PlanTypeEnum), nullable=False, default=PlanTypeEnum.GENERAL)

    # 状态信息
    status = Column(portable_enum(PlanStatusEnum), nullable=False, default=PlanStatusEnum.DRAFT)

    # 人员信息
    creator = Column(String(100), nullable=False)
    reviewer = Column(String(100))
    executors = Column(JSON, default=list)

    # 时间信息
    created_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    reviewed_at = Column(DateTime(timezone=True))
    started_at = Column(DateTime(timezone=True))
    completed_at = Column(DateTime(timezone=True))
    updated_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)

    # 计划内容
    steps = Column(JSON, nullable=False, default=list)
    estimated_total_hours = Column(Numeric(10, 2), default=0)
    actual_total_hours = Column(Numeric(10, 2))

    # 审核信息
    review_comment = Column(Text)
    rejection_reason = Column(Text)

    # 从属关系元数据
    is_based_on = Column(String(64), ForeignKey('plans.id', ondelete='SET NULL'))
    superseded_by = Column(String(64), ForeignKey('plans.id', ondelete='SET NULL'))

    # 扩展字段
    extra_metadata = Column('metadata', JSON, default=dict)

    # 关系
    task = relationship('Task', back_populates='plans', foreign_keys=[task_id])
    plan_steps = relationship('PlanStep', back_populates='plan', cascade='all, delete-orphan', order_by='PlanStep.step_index')

    # 约束
    __table_args__ = (
        CheckConstraint('version > 0', name='plans_version_check'),
        CheckConstraint('estimated_total_hours >= 0', name='plans_hours_check'),
        Index('idx_plans_task_id', 'task_id'),
        Index('idx_plans_status', 'status'),
        Index('idx_plans_type', 'type'),
        Index('idx_plans_creator', 'creator'),
        Index('idx_plans_reviewer', 'reviewer'),
        Index('idx_plans_created_at', 'created_at'),
        Index('idx_plans_reviewed_at', 'reviewed_at'),
        Index('idx_plans_version', 'task_id', 'version'),
    )

    def to_dict(self, include_task: bool = False) -> Dict[str, Any]:
        """转换为字典"""
        data = {
            'id': self.id,
            'task_id': self.task_id,
            'title': self.title,
            'version': self.version,
            'status': self.status.value if self.status else None,
            'type': self.type.value if self.type else None,
            'creator': self.creator,
            'reviewer': self.reviewer,
            'executors': self.executors or [],
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'reviewed_at': self.reviewed_at.isoformat() if self.reviewed_at else None,
            'started_at': self.started_at.isoformat() if self.started_at else None,
            'completed_at': self.completed_at.isoformat() if self.completed_at else None,
            'steps': self.steps or [],
            'estimated_total_hours': float(self.estimated_total_hours) if self.estimated_total_hours else 0,
            'actual_total_hours': float(self.actual_total_hours) if self.actual_total_hours else None,
            'review_comment': self.review_comment,
            'rejection_reason': self.rejection_reason,
            'is_based_on': self.is_based_on,
            'superseded_by': self.superseded_by,
            'metadata': self.extra_metadata or {},
        }

        if include_task and self.task:
            data['task'] = self.task.to_dict()

        return data


class PlanStep(Base):
    """
    计划步骤 (Plan Step)
    
    开发计划中的具体执行步骤
    """
    __tablename__ = 'plan_steps'

    id = Column(Integer, primary_key=True, autoincrement=True)
    plan_id = Column(String(64), ForeignKey('plans.id', ondelete='CASCADE'), nullable=False)
    step_index = Column(Integer, nullable=False)
    name = Column(String(200), nullable=False)
    description = Column(Text)
    estimated_hours = Column(Numeric(10, 2), nullable=False, default=0)
    status = Column(portable_enum(PlanStepStatusEnum), nullable=False, default=PlanStepStatusEnum.PENDING)
    executor = Column(String(100))
    started_at = Column(DateTime(timezone=True))
    completed_at = Column(DateTime(timezone=True))
    output = Column(Text)
    dependencies = Column(JSON, default=list)

    # 关系
    plan = relationship('DevelopmentPlan', back_populates='plan_steps')

    # 约束
    __table_args__ = (
        CheckConstraint('estimated_hours >= 0', name='plan_steps_hours_check'),
        Index('idx_plan_steps_plan_id', 'plan_id'),
        Index('idx_plan_steps_status', 'status'),
        Index('idx_plan_steps_executor', 'executor'),
    )

    def to_dict(self) -> Dict[str, Any]:
        """转换为字典"""
        return {
            'id': self.id,
            'plan_id': self.plan_id,
            'step_index': self.step_index,
            'name': self.name,
            'description': self.description,
            'estimated_hours': float(self.estimated_hours) if self.estimated_hours else 0,
            'status': self.status.value if self.status else None,
            'executor': self.executor,
            'started_at': self.started_at.isoformat() if self.started_at else None,
            'completed_at': self.completed_at.isoformat() if self.completed_at else None,
            'output': self.output,
            'dependencies': self.dependencies or [],
        }


class TaskPlanRelation(Base):
    """
    任务 - 计划关联表
    
    记录任务与所有计划的历史关联关系，支持审计和版本追溯
    """
    __tablename__ = 'task_plans'

    id = Column(Integer, primary_key=True, autoincrement=True)
    task_id = Column(String(32), ForeignKey('tasks.id', ondelete='CASCADE'), nullable=False)
    plan_id = Column(String(64), ForeignKey('plans.id', ondelete='CASCADE'), nullable=False)
    relation_type = Column(String(32), nullable=False, default='owns')
    is_active = Column(Boolean, nullable=False, default=False)
    activated_at = Column(DateTime(timezone=True))
    deactivated_at = Column(DateTime(timezone=True))
    created_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)

    # 约束
    __table_args__ = (
        Index('idx_task_plans_task_id', 'task_id'),
        Index('idx_task_plans_plan_id', 'plan_id'),
        Index('idx_task_plans_active', 'task_id', 'is_active'),
        Index('idx_task_plans_relation_type', 'relation_type'),
    )

    def to_dict(self) -> Dict[str, Any]:
        """转换为字典"""
        return {
            'id': self.id,
            'task_id': self.task_id,
            'plan_id': self.plan_id,
            'relation_type': self.relation_type,
            'is_active': self.is_active,
            'activated_at': self.activated_at.isoformat() if self.activated_at else None,
            'deactivated_at': self.deactivated_at.isoformat() if self.deactivated_at else None,
            'created_at': self.created_at.isoformat() if self.created_at else None,
        }
